File: algorithms/hard/longest_path_with_different_adjacent_characters.ex
defmodule Solution do
  @spec longest_path(parent :: [integer], s :: String.t()) :: integer
  def longest_path(parent, s) do
    s =
      s
      |> String.graphemes()
      |> Enum.with_index()
      |> Enum.reduce(%{}, fn {c, i}, acc ->
        Map.put(acc, i, c)
      end)

    parent
    |> Enum.with_index()
    |> Enum.reduce(%{}, fn {n, i}, acc ->
      Map.update(acc, n, [i], &[i | &1])
    end)
    |> dfs(0, s, 1)
    |> elem(1)
  end

  def dfs(mp, root, s, ans) do
    mp
    |> Map.get(root, [])
    |> Enum.reduce({0, 0, ans}, fn child, {m, n, ans} ->
      {ct, ans} = dfs(mp, child, s, ans)

      if s[child] != s[root] do
        [m, n] = Enum.sort([m, n, ct], &>=/2) |> Enum.take(2)
        {m, n, ans}
      else
        {m, n, ans}
      end
    end)
    |> then(fn {m, n, ans} ->
      ans = max(ans, m + n + 1)
      {m + 1, ans}
    end)
  end
end

File: algorithms/hard/longest_path_with_different_adjacent_characters.cpp
class Solution {
public:
  unordered_map<int, vector<int>> mp;
  string s;
  int ans = 1;
  int longestPath(vector<int> &parent, string s) {
    for (int i = 0; i < parent.size(); i++) {
      mp[parent[i]].push_back(i);
    }
    this->s = s;
    dfs(0);
    return ans;
  }
  int dfs(int root) {
    int m = 0, n = 0;
    for (auto child : mp[root]) {
      int ct = dfs(child);
      if (s[child] != s[root]) {
        if (ct > n) n = ct;
        if (n > m) swap(m, n);
      }
    }

    ans = max(ans, m + n + 1);
    return m + 1;
  }
};
